### ISSUE TITLE
[research] Masked-diffusion training objective + data collation (portable)

### ISSUE THREAD

**Issue by travisgalloway** [travisgalloway/monica#44]
Part of #42. Depends on #D1 (bidirectional backbone).

**Goal:** The portable masked-denoising objective that replaces next-token cross-entropy for the diffusion track. Absorbing-state diffusion, LLaDA-style: sample a mask ratio `t ~ U(0,1]`, replace the selected tokens with a `[MASK]` id, predict the **masked positions only**, and weight the loss by `1/t` (the absorbing-diffusion ELBO weighting). The training loop (`src/train/loop.py`) is already backend-agnostic and injects `train_step`, so only the objective + collation + a backend value_and_grad wrapper change — not the loop.

**Acceptance criteria**
- [ ] New portable `src/train/diffusion_objective.py` (no `mlx`/`torch` import) providing: mask-ratio sampling, the `[MASK]`-replacement collation, and the masked-only `1/t`-weighted cross-entropy. Added to `PORTABLE_MODULES` in `tests/test_import_guard.py`.
- [ ] `[MASK]` token handled by `vocab_size += 1`; **re-check `MambaConfig.validate()`'s `< 65536` uint16 invariant** (poc vocab 50280 + 1 is fine).
- [ ] A collation path so the loader stops shifting targets by one for this track (today `src/data/loader.py::_collate` returns `arr[:, :-1], arr[:, 1:]`).
- [ ] Unit tests: mask-ratio sampling distribution, masked-only CE (unmasked positions contribute zero), and the `1/t` weighting.
- [ ] A backend `make_train_step` variant wiring this loss decreases loss on toy data.

**Files:** `src/train/diffusion_objective.py` (new, portable), `src/data/loader.py` (collation helper), `src/model/mlx_train_step.py` (loss variant), `tests/test_import_guard.py`, `tests/test_diffusion_objective.py` (new).

Seam note: objective + collation are portable; only the `nn.value_and_grad` wrapper is backend.

## Verification
1. `.venv/bin/python -m pytest tests/test_diffusion_objective.py tests/test_import_guard.py`
2. Toy diffusion train_step over a few steps: loss decreases (covered fully by #D5).

**Comment by travisgalloway** [comment]
Consolidated into and owned by #203 ('[M12] P4-P5: diffusion arm + LSP discriminator'), which absorbs the discrete-diffusion research track. Closed during closure audit to groom the backlog.
